classes/local/step/connector_curl.php

```diff
@@ -27,18 +27,181 @@
+            $errors['config_rawpostdata'] = get_string('config_field_missing', 'tool_dataflows', 'rawpostdata', true);
+        }
+        return empty($errors) ? true : $errors;
+    }
+
     /**
      * Executes the step
      *
-     * TODO: This will perform a curl call
+     * Performs a curl call according to given parameters.
      *
      * @return bool Returns true if successful, false otherwise.
      */
     public function execute(): bool {
-        // TODO: Implement.
+        global $CFG;
+        // Get variables.
+        $config = $this->enginestep->stepdef->config;
+        $isdryrun = $this->enginestep->engine->isdryrun;
+        $method = $config->method;
+        $dbgcommand = 'curl -X ' .  strtoupper($method) . ' ' . $config->curl;
+        $result = null;
+
+        if (!empty($config->timeout)) {
+            $this->timeout = (int) $config->timeout;
+            $dbgcommand .= ' --max-time ' . $config->timeout;
+        }
+
+        $options = ['CURLOPT_TIMEOUT' => $this->timeout];
+
+        if ($method == 'post') {
+            $options['CURLOPT_POST'] = 1;
+        }
+
+        if ($method == 'put') {
+            $options['CURLOPT_PUT'] = 1;
+        }
+
+        $curl = new \curl();
+
+        if (!empty($config->sideeffects)) {
+            $this->hassideeffect = true;
+        } else {
+            $this->hassideeffect = false;
+        }
+
+        // Provided a header is specified add header to request.
+        if (!empty($config->headers)) {
+            $headers = $config->headers;
+            if (!is_array($headers)) {
+                $headers = json_decode($headers, true);
+            }
+            $curl->setHeader($headers);
+            $dbgcommand .= ' -H \'' . $config->headers . '\'';
+        }
+
+        // Sets post data.
+        if (!empty($config->rawpostdata)) {
+            $options['CURLOPT_POSTFIELDS'] = $config->rawpostdata;
+            $dbgcommand .= ' -d \'' . $config->rawpostdata . '\'';
+        }
+
+        // Download response to file provided destination is set.
+        if (!empty($config->destination) && !$isdryrun) {
+            if ($config->destination[0] === '/') {
+                $config->destination = substr($config->destination, 1);
+            }
+            $config->destination = $this->enginestep->engine->resolve_path($config->destination);
+            $file = fopen($config->destination, 'w');
+            $options['CURLOPT_FILE'] = $file;
+        }
+
+        // Perform call.
+        if (!$isdryrun) {
+            $result = $curl->$method($config->curl, [], $options);
+            // For put and post methods automatically overrides/has side effects.
+            if ($config->method != 'get') {
+                $this->hassideeffect = true;
+            }
+        }
+
+        if (!empty($config->destination) && !$isdryrun) {
+            fclose($file);
+        }
```

Comment:
I think this can simply be 
```suggestion
        if (!empty($file)) {
            fclose($file);
        }
```